index.rst

```diff
@@ -1,61 +1,499 @@
+
+- like ``Datastore``, this aspect would be backed by files in a shared filesystem or object store (but not necessarily the same filesystem or bucket as an associated ``Datastore``);
+- like ``Registry``, the new aspect would provide descriptive and organizational metadata for datasets, rather than hold datasets themselves, and after execution its content would be completely loaded into the ``Registry``.
+
+The full high-level design will be the subject of a future technote, but an early sketch can be found `in Confluence <https://confluence.lsstcorp.org/display/DM/Saving+per-Quantum+provenance+and+propagating+nothing-to-do+cases%2C+and+The+Future>`_.
+A key point is that ultimately we want inserting datasets back into a permanent registry to be a deferred, final step in all modes of execution, not just BPS, allowing that step to transform file-based data in various ways before it is loaded into a database.
+
+Recording provenance only when using BPS (and relying on it to manage the ``QuantumGraph`` and provenance files) in the interim seems like a good first step.
+Extending the design to include non-BPS processing may take time, but we do not anticipate it changing what happens at a low level, or the appearance of persisted provenance information.
+
+.. _failures_and_retries:
+
+Failures and retries
+--------------------
+
+When a quantum fails during execution, we will often want to try running it again.
+This may happen immediately via workflow system functionality or manually via a resubmission.
+The outputs of any such failure are considered unreliable, and we don't want to proceed with downstream processing even if some outputs were written before the failure (not producing all - or even any - outputs is not necessarily considered a failure).
+We want actual failures to be included in provenance, and their datasets included in the main data repository, until/unless they are explicitly deleted after any analysis of the failure is complete.
+This leads to a conflict: running a failed quantum again would lead to two ``quantum`` rows with the same UUID, data ID, and ``RUN``, and analoguous conflicts for any output datasets that were written before the failure occurred or in spite of it (e.g. logs written by the execution system after it catches an exception).
+
+The possibility of immediate retries performed by the workflow system means we cannot assign the retry quantum and datasets new UUIDs, because this would prevent downstream quanta from being able to find their inputs.
+Instead, we have to assign new UUIDs (and a new RUN collection) to the failures, and move all datasets that were already written to permanent storage out of the way.
+This could be done immediately after the failure occurs if we have a process able to do the work (i.e. the worker process fails gracefully, or some control process is able to take action), or it could be done only just before actually attempting a retry.
```

Comment:
The safest would be before actually attempting a retry (e.g., preemption on the IDF).  Is everything in a recoverable state if something happens between moving some files and actually rerunning the Quanta?

Reply:
Agree that it's safest to do right before attempting a retry, so we should always do it there, but I also think we always need to do it in the transfer job for cases that haven't been retried, to avoid putting non-permanent locations in (e.g.) Rucio and on-permanent UUIDs in the main registry when at all possible.

I think it's all recoverable if the logic responsible for this "move out of the way" operation is written carefully:
- write the updated UUIDs and their connections to a new Quantum file in a deterministic location (or at least a location we can quickly search for the existence of such a file);
- only move other artifacts to that location after it has been written.